src/vk/vk-contacts/vkcontactsyncadaptor.cpp

```diff
@@ -0,0 +1,1010 @@
+                        sHome = phn;
+                    }
+                }
+                if (mMobile.number() != sMobile.number()) {
+                    mMobile.setNumber(sMobile.number());
+                    saveNonexportableDetail(modified, mMobile);
+                    *hasChanged = true;
+                }
+                if (mHome.number() != sHome.number()) {
+                    mHome.setNumber(sHome.number());
+                    saveNonexportableDetail(modified, mHome);
+                    *hasChanged = true;
+                }
+                return i;
+            }
+        }
+
+        // the contact doesn't exist in the local database.
+        // it must be a remote addition.
+        return -1;
+    }
+
+    void calculateRemoteDelta(const QList<QContact> &allReceivedContacts, const QList<QContact> &localVkContacts, QList<QContact> *additions, QList<QContact> *modifications, QList<QContact> *deletions)
+    {
+        // determine deletions
+        for (int i = 0; i < localVkContacts.size(); ++i) {
+            const QContact &c(localVkContacts[i]);
+            const QString &existingGuid = c.detail<QContactGuid>().guid();
+            bool found = false;
+            foreach (const QContact &rc, allReceivedContacts) {
+                if (rc.detail<QContactGuid>().guid() == existingGuid) {
+                    found = true;
+                    break;
+                }
+            }
+            if (!found) {
+                // must have been deleted remotely.
+                deletions->append(c);
+            }
+        }
+
+        // determine additions/modifications
+        for (int i = 0; i < allReceivedContacts.size(); ++i) {
+            QContact c = allReceivedContacts[i];
+            bool hasChanged = false;
+            int idx = matchingContactIndex(&c, localVkContacts, &hasChanged);
+            if (idx < 0) {
+                // new contact
+                additions->append(c);
+            } else if (hasChanged) {
+                // modified contact
+                modifications->append(c);
+            }
+        }
+    }
+}
+
+VKContactSyncAdaptor::VKContactSyncAdaptor(QObject *parent)
+    : VKDataTypeSyncAdaptor(SocialNetworkSyncAdaptor::Contacts, parent)
+    , QtContactsSqliteExtensions::TwoWayContactSyncAdapter(QStringLiteral("vk"))
+    , m_workerObject(new VKContactImageDownloader())
+{
+    connect(m_workerObject, &AbstractImageDownloader::imageDownloaded,
+            this, &VKContactSyncAdaptor::imageDownloaded);
+
+    // can sync, enabled
+    setInitialActive(true);
+}
+
+VKContactSyncAdaptor::~VKContactSyncAdaptor()
+{
+    delete m_workerObject;
+}
+
+QString VKContactSyncAdaptor::syncServiceName() const
+{
+    return QStringLiteral("vk-contacts");
+}
+
+void VKContactSyncAdaptor::sync(const QString &dataTypeString, int accountId)
+{
+    m_apiRequestsRemaining[accountId] = 99; // assume we can make up to 99 requests per sync, before being throttled.
```

Comment:
The API documentation says two per second and 5000 per hour https://vk.com/pages?oid=-17680044&p=Limitations_for_Calling_Ads_API_Methods